In [1]:
import pandas as pd
import optuna


from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import precision_score, recall_score, f1_score, classification_report, accuracy_score, confusion_matrix


In [2]:
X_train = pd.read_csv("/home/pavithra/projects/spam_or_ham_fsec/data/target/X_train.csv", header=None)
X_test = pd.read_csv("/home/pavithra/projects/spam_or_ham_fsec/data/target/X_test.csv", header=None)
y_train = pd.read_csv("/home/pavithra/projects/spam_or_ham_fsec/data/target/y_train.csv", header=None)
y_test = pd.read_csv("/home/pavithra/projects/spam_or_ham_fsec/data/target/y_test.csv", header=None)



print("X train ",X_train.shape)
print("X test",X_test.shape)
print("Y_Train", y_train.shape)
print("Y_test", y_test.shape)

X train  (4050, 1000)
X test (375, 1000)
Y_Train (4050, 1)
Y_test (375, 1)


In [3]:
def objective(trial):
    n_estimators = trial.suggest_int('n_estimators', 50, 100)
    max_depth = trial.suggest_int('max_depth', 2, 32, log=True)
    min_samples_split = trial.suggest_int('min_samples_split', 2, 20)
    min_samples_leaf = trial.suggest_int('min_samples_leaf', 1, 20)
    max_features = trial.suggest_categorical('max_features', ['sqrt', 'log2', None])

    model = RandomForestClassifier(
        n_estimators=n_estimators,
        max_depth=max_depth,
        min_samples_split=min_samples_split,
        min_samples_leaf=min_samples_leaf,
        max_features=max_features,
        random_state=42,
        class_weight="balanced"
    )

    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)
    precision = f1_score(y_test, y_pred)

    return precision

study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=50)

print(f"Best hyperparameters: {study.best_params}")
print(f"Best precision: {study.best_value}")

[I 2024-12-02 00:54:58,631] A new study created in memory with name: no-name-8fff5498-fe66-47d5-92e8-c60e0c4c2f61
/home/pavithra/.local/lib/python3.8/site-packages/sklearn/base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
[I 2024-12-02 00:55:08,605] Trial 0 finished with value: 0.0 and parameters: {'n_estimators': 83, 'max_depth': 19, 'min_samples_split': 19, 'min_samples_leaf': 20, 'max_features': 'sqrt'}. Best is trial 0 with value: 0.0.
/home/pavithra/.local/lib/python3.8/site-packages/sklearn/base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
[I 2024-12-02 00:55:09,827] Trial 1 finished with value: 0.1686746987951807 and parameters: {'n_estimators': 75, '

Best hyperparameters: {'n_estimators': 62, 'max_depth': 3, 'min_samples_split': 5, 'min_samples_leaf': 19, 'max_features': 'sqrt'}
Best precision: 0.30508474576271183


Best hyperparameters: {'n_estimators': 62, 'max_depth': 3, 'min_samples_split': 5, 'min_samples_leaf': 19, 'max_features': 'sqrt'}
Best f1-score: 0.30508474576271183

In [4]:
param = {'n_estimators': 62, 'max_depth': 3, 'min_samples_split': 5, 'min_samples_leaf': 19, 'max_features': 'sqrt'}
model = RandomForestClassifier(
    **param,
    random_state=42
)

model.fit(X_train, y_train)

y_pred = model.predict(X_test)
precision = precision_score(y_test, y_pred)

/home/pavithra/.local/lib/python3.8/site-packages/sklearn/base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
/home/pavithra/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [5]:

def classification_metrics(y_test, y_pred):
    print("="*50)
    print("The accuracy score", accuracy_score(y_test, y_pred))
    print("Precision:", precision_score(y_test, y_pred))
    print("Recall:", recall_score(y_test, y_pred))
    print("F1 score:", f1_score(y_test, y_pred))
    print("="*50)
    print(classification_report(y_test, y_pred))
    print("="*50)
    print("confusion Matrix")
    print(confusion_matrix(y_test, y_pred))
    print("="*50)

classification_metrics(y_test, y_pred)

The accuracy score 0.8986666666666666
Precision: 0.0
Recall: 0.0
F1 score: 0.0
              precision    recall  f1-score   support

           0       0.90      1.00      0.95       337
           1       0.00      0.00      0.00        38

    accuracy                           0.90       375
   macro avg       0.45      0.50      0.47       375
weighted avg       0.81      0.90      0.85       375

confusion Matrix
[[337   0]
 [ 38   0]]


/home/pavithra/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/pavithra/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/pavithra/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/pavithra/.local/lib/python3.8/site-packages/sklea